feat: `optional` codec and data type

LDeakin opened the pull request:
I'm still finalising an implementation, but here is a draft spec.

LDeakin committed:
feat: `optional` codec and data type

jbms commented:
Looks good, the only issue I see is the fill value representation not allowing null for the base data type fill value, e.g. for a base data type of json or nested optional.  Instead you could specify the base data type fill value in a single-element array:

null -> missing 
[1] -> value of 1
[null] -> value of null

LDeakin committed:
fix: remove flatten from mask

LDeakin commented:
I've implemented this in `zarrs` with codec/data type support for arbitrarily nested optional data. I'll add example data later. I have not addressed these two issues:

> the only issue I see is the fill value representation not allowing null for the base data type fill value

> for a multiply nested optional type like Option<Option<u8>>, you could potentially have the "null" at different levels

I am open to explicit suggestions that satisfy both, or only the first. The latter is a bit burdensome to support for something I suspect nobody would use. What I currently do:
-  `null` always means an empty element with the `optional` data type, irrespective of the inner type
- A non-null fill value gets propagated down to the deepest non-optional inner type.

mkitti commented:
The other two null-like types I would immediately like to use this for are:
1. [`NA`](https://cran.r-project.org/doc/manuals/r-release/R-lang.html#NA-handling) or [`missing`](https://docs.julialang.org/en/v1/manual/missing/) to represent missing data in the statistical sense.
2. `NaN` as a general analog to IEEE 754 floating point. Practically we probably just using a floating point type here.

I do not think these are well represented by `null`. Each has their own semantics.

```julia
julia> NaN == true
false

julia> missing == true
missing
```

If we could generalize this to any two types and introduce a `null_type`, I think this could cover a much wider array of statistical and numerical applications.

LDeakin committed:
feat: change `optional` data type inner fill value representation

LDeakin committed:
fix: revise `optional` JSON examples

LDeakin committed:
feat: add `optional` example data

LDeakin commented:
> I previously suggested wrapping any non-None fill value in a one-element array. That solves both issues and is syntactically pretty minimal.

Thanks @jbms, that ended up working out quite nicely! I've updated the spec and implementation, and added example data.

LDeakin commented:
No problem, `zarr-python` could consider `StringDType(na_object=nan)` as mapping to
```json
 "data_type": {
    "name": "optional",
    "configuration": {
      "name": "string",
      "configuration": {}
    }
```
In zarr.json, and handle converting to a masked representation internally. `null`s don't need to be serialised with the actual string data, the mask is used to reconstruct them.

flying-sheep commented:
OK, let me rephrase: `np.nan` isn’t special, it just happens to be a common choice for a “NA object” that is already serializable by zarr-python (as a 1-element float array).

I’m questioning this:

> The value of the `fill_value` metadata key MUST be `null` or a single element array containing any valid fill value of the underlying data type.

In the Python world, the NA objects are often *not necessarily* of the same type as the array: A nullable numpy string array dtype might have it set to `""` (`str`), `None` (`types.NoneType`), `np.nan` (`float`), or `pandas.NA` (???).

So I guess as given, there would be no way to express that, and it would have to be expressed externally (e.g. by storing a separate 1 element array with the NA object.

Would it make sense to have this NA value be different from a `fill_value` (as used in other parts of zarr) to allow for objects that aren’t of the same type, or should this spec stay as it is (which means that plain zarr-python probably won’t support nullable string arrays that don’t happen to have a `str`-typed NA value)

mkitti reviewed (commented):
We may need to define `null` better here. Perhaps we should follow `null` from Polars and Apache  Arrow.

LDeakin commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,80 @@
+
+For nested optional types, this representation is applied recursively.
+
+The table below demonstrates valid `data_type` and `fill_value` combinations with an `optional` and nested `optional` data type, along with their equivalent Rust [`Option`](https://doc.rust-lang.org/std/option/) values.
```
> It may be necessary to define `null` as its own data type

Why? `null` is just being used as a fill value with specific behaviour for the `optional` data type.

> From the scale-offset codec conversation

Codecs like `scale_offset` could add support for the `optional` data type if there is demand, but I think that is outside the scope of this PR. Inner codecs in `data_codecs` would not interact with the `optional` data type anyway. See below:

> While array-to-array codecs MAY support the `optional` data type, implementations SHOULD use the `optional` codec as the sole top-level codec.
> This approach is preferred because the codecs contained within the `optional` codec configuration do not need to explicitly handle optional data type semantics.

Array-to-array codecs that perform shape manipulation (e.g. [`reshape`](https://github.com/zarr-developers/zarr-extensions/pull/10)) could be an exception here as they support all data types.

mkitti replied:
> Why? `null` is just being used as a fill value with specific behaviour for the `optional` data type.

`null` has meaning in other contexts, particularly in other storage contexts. The term arises in SQL, Apache Arrow, and Polars. In the data type fill value description here, we state that `null` indicates the absence of value.

There are two possible ways to interpret an absence of value:
1. The value does not exist.
2. The value is not known.

The first interpretation is typically from the perspective of a software engineer. The value is completely absent. This may be because it was not defined. Trying to access a value that was has not been defined is then thought of as an error that must be handled.

The second interpretation may be from a statistican. The value is missing because we have not measured it or we do not know it. The value may exist, but it is not known to us. Trying to access the value is not an error but rather the lack of knowledge is represented.

Knowing the history of this specification, I would guess that you may mean the first interpretation. In Zarr v2, we also a null fill value meant "no fill value" or the values of an "empty chunk" are undefined. However, in the context of other analogous data libraries, SQL, Apache Arrow, and Polars, `null` means the second interpretation.

Please clarify how `null` should be interpreted here. Leaving the interpretation ambiguous will result in interoperability issues.

mkitti replied:
> I don't think zarr usually defines runtime behavior of data, which would mean that discussing the “best” semantics here is probably a distraction. That being said numpy explicitly made it possible to use e.g. `pd.NA` there so you're treating it too harshly.

The issue I'm pointing out is that datatypes do come with some implied runtime behavior. The floating point types rely on IEEE 754 for both their formatting, but this also describes arithmetic on those types. There is implied modular arithmetic on signed and unsigned integers of fixed width.

In this case, the definition of this codec and data type come very close to how [Pola.rs](https://pola.rs/) adresses missing data. For each column, Pola.rs also stores a validity array. In Rust, missing values are also referred to by `Option<T>` when interacting with `Series`.

```rust
use polars::prelude::*;

fn main() -> PolarsResult<()> {
    // Explicitly defining a vector with missing values
    let values: Vec<Option<i32>> = vec![
        Some(10), 
        None,      // This becomes a null in Polars
        Some(30), 
        None
    ];

    let s = Series::new("counts".into(), values);
    
    println!("Series with nulls:\n{}", s);
    Ok(())
}

fn decode_series(s: &Series) {
    // Convert a Series back into a Vec of Options
    // .i32() attempts to view the series as Int32 chunks
    let decoded: Vec<Option<i32>> = s.i32()
        .unwrap()
        .into_iter()
        .collect();

    for val in decoded {
        match val {
            Some(v) => println!("Found value: {}", v),
            None    => println!("Found a missing entry!"),
        }
    }
}
```

However, in Pola.rs,  you can multiply a column with missing data by a scalar.

```rust
use polars::prelude::*;

fn main() -> PolarsResult<()> {
    // 1. Create a DataFrame with Some values and None (nulls)
    let df = df!(
        "sensor_id" => ["A1", "A2", "A3", "A4"],
        "reading" => [Some(10.0), None, Some(25.0), None]
    )?;

    // 2. Multiply the "reading" column by a scalar (e.g., a calibration factor of 1.5)
    // Polars handles the 'None' entries automatically—they stay 'null'.
    let calibrated_df = df.lazy()
        .with_column(
            (col("reading") * lit(1.5)).alias("calibrated_reading")
        )
        .collect()?;

    println!("Resulting DataFrame:\n{}", calibrated_df);

    Ok(())
}
```

This results in the following output:

```
Resulting DataFrame:
shape: (4, 3)
┌───────────┬─────────┬────────────────────┐
│ sensor_id ┆ reading ┆ calibrated_reading │
│ ---       ┆ ---     ┆ ---                │
│ str       ┆ f64     ┆ f64                │
╞═══════════╪═════════╪════════════════════╡
│ A1        ┆ 10.0    ┆ 15.0               │
│ A2        ┆ null    ┆ null               │
│ A3        ┆ 25.0    ┆ 37.5               │
│ A4        ┆ null    ┆ null               │
└───────────┴─────────┴────────────────────┘
```

flying-sheep replied:
Apache Arrow defines null as “unknown”, not “missing” https://arrow.apache.org/blog/2022/10/05/arrow-parquet-encoding-part-1/

LDeakin committed:
fix: clarify that the `name` is Zarr V3

LDeakin committed:
fix: formalise that the `configuration` field is optional

LDeakin committed:
fix: add `undefined` to the mix

LDeakin committed:
fix: codec clarifications, recommendations for scale-offset etc

LDeakin committed:
fix: clarifications on missing value interpretation and in-memory representations

LDeakin commented:
@mkitti @flying-sheep  etc, I have made some changes that hopefully address concerns.

Firstly, I've written a fairly general section on how other array-to-array codecs could deal with optional data types.

Secondly, I've added a recommendation that if implementations wish to impose a specific in-memory representation for null fill values, they should do that through a registered attribute. I think this is fairly reasonable as opposed to adding an additional field to `optional` data type, since it does not impact encoding/decoding and it would be ignored by implementations in alternate languages. I just don't like the idea of mixing in language-specific implementation details that do not impact the actual encoding and decoding of the data.

Spitballing (I have no intention of standardising this or anything similar myself):
```json
"attributes": {
  "py_array_representation": "np.ma.MaskedArray"
  # "py_array_representation": "np.typing.NDArray[np.dtypes.StringDType(na_object=np.nan)]"
  "julia_optional_data_type_missing_element_representation": "missing"
}
```
There could even be a more targeted convention around how missing data should be interpreted that is language-agnostic (e.g. missing, undefined, unknown). But I'd still recommend that as an attribute, and it should not block this PR.

mkitti commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,116 @@
+
+### Julia
+
+The Julia `missing` or `NA` values could be used to represent null values in an `optional` data type.
```
To clarify, the Julia types are [`missing`](https://docs.julialang.org/en/v1/manual/missing/) or [`nothing`](https://docs.julialang.org/en/v1/base/constants/#Core.nothing).

* `missing` is the statistician's null. You can multiply it and perform ternary (Kleene) logic with it.
* `nothing` is the programmer's null or void. It does not participate in computation. It is used similar to `None` in Python.

`NA` is used in the R programming language and is the equivalent of `missing` in Julia.

LDeakin committed:
fix: Julia `nothing`

LDeakin committed:
fix: full Julia optional type

mkitti commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,117 @@
+
+The `null` value type could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.
+An `enum`-like codec would be required to distinguish between `missing` or `NA` values within a single array.
+
```
```suggestion
### R

In R, `NA` may be used to represent `null` values in an optional data type.
```

LDeakin committed:
fix: mention ternay logic

LDeakin committed:
feat: mention R

LDeakin committed:
fix: remove enum comment

LDeakin committed:
fix: reword ternary logic sentence

mkitti commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,117 @@
+
+A `numpy` array using the `StringDType` with an `na_object` that is not `None` could use the `optional` data type with a `string` underlying data type.
+However, the `na_object` itself would not be stored in the Zarr metadata of the `optional` data type.
+The `na_object` could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.
```
```suggestion
In Python, the representation of `null` values in an `optional` data type may depend on the use of other libraries.
* `None` may be used when using only the Python standard library.
* [`pandas.NA`](https://pandas.pydata.org/docs/reference/api/pandas.NA.html) may be used in conjunction with Pandas.
* [`polars.null`](https://docs.pola.rs/user-guide/expressions/missing-data/) is an appropriate direct mapping for Polars.
* [`pyarrow.null`](https://arrow.apache.org/docs/python/generated/pyarrow.null.html#pyarrow.null) could also be used with pyarrow.
* [`np.nan`](https://numpy.org/doc/2.3/reference/constants.html#numpy.nan) or [`numpy.ma.MaskedArray`](https://numpy.org/doc/stable/reference/maskedarray.generic.html) could be used in NumPy.

A `numpy` array using the `StringDType` with an `na_object` that is not `None` could use the `optional` data type with a `string` underlying data type.
However, the `na_object` itself would not be stored in the Zarr metadata of the `optional` data type.
The `na_object` could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.
```

mkitti commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,122 @@
+
+Defines a data type for optional (nullable) values that can contain either a value of a specified underlying data type or be missing/undefined/null.
+
+The `optional` data type does not define how null values should be _interpreted_ and represented in-memory.
```
```suggestion
The `optional` data type does not define how null values should be _interpreted_, used semantically, or represented in-memory.
```

LDeakin commented on data-types/optional/README.md:
```diff
@@ -0,0 +1,122 @@
+
+For nested optional types, this representation is applied recursively.
+
+## In-memory representations
```
It is invalid until it is spec'd. A `bytes` representation would need to be defined for optional data.

I'd rather not include it here and would encourage the split data / mask encoding. However, the `bytes` codec could be supported simply with a tag (present/missing) + payload encoding, always the same size. A more efficient Rust-like [niche optimisation](https://jpfennell.com/posts/enum-type-size/) could be supported for certain inner types, but perhaps that is a bit too fancy?
